src/pcap/client_test.go

```diff
@@ -0,0 +1,287 @@
+	0x052, 0x065, 0x066, 0x065, 0x072, 0x065, 0x072, 0x03a, 0x020, 0x068, 0x074, 0x074, 0x070, 0x03a, 0x02f, 0x02f,
+	0x077, 0x077, 0x077, 0x02e, 0x065, 0x074, 0x068, 0x065, 0x072, 0x065, 0x061, 0x06c, 0x02e, 0x063, 0x06f, 0x06d,
+	0x02f, 0x064, 0x065, 0x076, 0x065, 0x06c, 0x06f, 0x070, 0x06d, 0x065, 0x06e, 0x074, 0x02e, 0x068, 0x074, 0x06d,
+	0x06c, 0x00d, 0x00a, 0x00d, 0x00a, 0x0}
+
+type MessageTuple struct {
+	Response *CaptureResponse
+	err      error
+}
+
+type MockAPIWriter struct {
+	index    int
+	messages []MessageTuple
+	grpc.ClientStream
+}
+
+func (x *MockAPIWriter) Send(m *CaptureRequest) error {
+	return x.ClientStream.SendMsg(m)
+}
+
+func (x *MockAPIWriter) Recv() (*CaptureResponse, error) {
+	if x.index >= len(x.messages) {
+		return nil, io.EOF
+	}
+	response := x.messages[x.index].Response
+	err := x.messages[x.index].err
+	x.index++
+	return response, err
+}
+
+func TestMain(m *testing.M) {
+	// setup logger
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		panic(err)
+	}
+	zap.ReplaceGlobals(logger)
+	os.Exit(m.Run())
+}
+
+func TestHandleStream(t *testing.T) {
+	tests := []struct {
+		name               string
+		clientError        error
+		messages           []MessageTuple
+		expectedErrMessage string // TODO: better way to compare errors
+	}{
+		{
+			name:        "clean stop",
+			clientError: nil,
+			messages: []MessageTuple{
+				{
+					Response: newPacketResponse(examplePacket, gopacket.CaptureInfo{}),
+				},
+			},
+			expectedErrMessage: "context canceled",
+		}, {
+			name:        "client stop",
+			clientError: fmt.Errorf("client stop"),
+			messages: []MessageTuple{
+				{
+					Response: newPacketResponse(examplePacket, gopacket.CaptureInfo{}),
+				},
+			},
+			expectedErrMessage: "client stop",
+		}, {
+			name:        "Permission Denied",
+			clientError: nil,
+			messages: []MessageTuple{
+				{
+					err: errorf(codes.PermissionDenied, "error-text"),
+				},
+			},
+			expectedErrMessage: "receive non-OK code: PermissionDenied: error-text",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var buf bytes.Buffer
+			writer := pcapgo.NewWriter(&buf)
+			copyWg := &sync.WaitGroup{}
+			copyWg.Add(1)
+			stream := &MockAPIWriter{messages: tt.messages}
+			ctx, cancel := WithCancelCause(context.Background())
+			go handleStream(stream, writer, copyWg, cancel)
+			if tt.clientError != nil {
+				cancel(tt.clientError)
+			}
+			copyWg.Wait()
+
+			err := Cause(ctx)
+			if err.Error() != tt.expectedErrMessage {
+				t.Errorf("expected = %v, actual = %v", tt.expectedErrMessage, err)
+			}
+		})
+	}
+}
+
+func TestNewClient(t *testing.T) {
+	// remove test file if already present
+	outputFile := "testpcapfile.pcap"
+	_ = os.Remove(outputFile)
+
+	tests := []struct {
+		name              string
+		outputFile        string
+		wantErr           bool
+		expectedErrString string
+	}{
+		{
+			name:       "first run - file does not exist",
```

Comment:
```suggestion
			name:       "output file does not exist",
```

1. there should not be a deterministic order for test cases.
2. The label 'first run' does not add any information.